### ISSUE TITLE
"collectd_iostat_python" cannot be loaded

### ISSUE THREAD

**Issue by saina110** [deniszh/collectd-iostat-python#4]
Dear all,

Recently, I have tried to make a custom plugin to collect data data from the system. The platform I am using is **Solaris 10 Sparc** 

I am using "collectd_iostat_python" plugin which is made by [Mr.deniszh](https://github.com/deniszh/collectd-iostat-python), but when I try to run the collectd, I found the issues in the log file:

> **python plugin: Found a configuration for the "collectd_iostat_python" plugin, but the plugin isn't loaded or didn't register a configuration callback.**

And the collectd will automatically stop running after 10 seconds. 

_the collectd.conf file_:

```
<LoadPlugin python>
  Globals true
 </LoadPlugin>
 ....
 <Plugin python>
    ModulePath "/home/yuanliao/Documents/tmp"
    Import "collectd_iostat_python"

    <Module collectd_iostat_python>
        Path "/usr/bin/iostat"
                Interval 2
                Count 2
                Verbose False
    </Module>
 </Plugin>
```

While in the collectd_iostat_python.py file, the modifications I made is according to the deniszh's proto.

This issue confused me a long time. But I dont get some useful clue about it. 

I would be really appreciated for your help and suggestions. Thanks in advance. 

Regards

**Comment by deniszh** [comment]
Oops... Solaris sparc... Sorry, have no access to that to check and fix.
Could you please show what located in "/home/yuanliao/Documents/tmp" directory?
It's need to be only collectd_iostat_python.py file in that specific dir, not in subdirs.

**Comment by deniszh** [comment]
Please also check if 'iostat' is really located in /usr/bin on Solaris (I'm doubt).
And please also run 'iostat -kNd' and 'iostat -kNdx' and send me output - if it works on Solaris, of course.

**Comment by saina110** [comment]
Hi deniszh,

Thank you for your suggestions. I just tried the methods you recommended. 

> 1) in the "/home/yuanliao/Documents/tmp" , I remove the other files and keep the collectd_iostat_python.py in it, not in the subdirs. But it shows the same error.
> 
> 2) 'iostat' is located in /usr/bin on my system, while there is no commend **iostat -kNd**, so I use **iostat -xtc** to replace. When I type **iostat -xtc** in terminal, the output is below:
> 
> ```
> extended device statistics                     tty         cpu
> device     r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b  tin tout  us sy wt id
> sd3        0.0    0.0    0.0    0.0  0.0  0.0    0.9   0   0    0   43   1  1  0 98
> sd4        0.6    0.4    2.4    2.5  0.0  0.0   22.3   0   1 
> sd5        1.7    3.1   10.5   25.2  0.0  0.1   29.7   0   3 
> sd6        0.2    0.0    9.5    0.0  0.0  0.0    4.4   0   0 
> sd7        0.5    1.0    2.1    5.4  0.0  0.0   31.0   0   1 
> nfs1       0.0    0.0    0.0    0.0  0.0  0.0    0.0   0   0 
> nfs4055    0.0    0.0    0.0    0.0  0.0  0.0    5.3   0   0 
> nfs4212    2.0    3.4   11.6  108.2  0.3  0.0   58.8   0   0 
> nfs4228    0.0    0.0    0.0    0.0  0.0  0.0    0.0   0   0 
> nfs4229    0.0    0.0    0.1    0.0  0.0  0.0    1.0   0   0 
> nfs4230    0.0    0.0    0.1    0.0  0.0  0.0    0.8   0   0 
> nfs4231    0.0    0.0    0.0    0.0  0.0  0.0    0.8   0   0 
> ```
> 
> when I type **python collectd_iostat_python.py**, the output is:
> 
> ```
> sd4.svc_t:0.0
> sd4.w_s:0.0
> sd4.actv:0.0
> sd4._b:0.0
> sd4.kw_s:0.0
> sd4.kr_s:0.0
> sd4._w:0.0
> sd4.r_s:0.0
> sd4.wait:0.0
> sd5.svc_t:30.5
> sd5.w_s:31.4
> sd5.actv:1.0
> sd5._b:16.0
> sd5.kw_s:76.6
> sd5.kr_s:0.0
> sd5._w:0.0
> sd5.r_s:0.0
> sd5.wait:0.0
> sd6.svc_t:0.0
> sd6.w_s:0.0
> sd6.actv:0.0
> sd6._b:0.0
> sd6.kw_s:0.0
> sd6.kr_s:0.0
> sd6._w:0.0
> sd6.r_s:0.0
> sd6.wait:0.0
> sd7.svc_t:0.0
> sd7.w_s:0.0
> sd7.actv:0.0
> sd7._b:0.0
> sd7.kw_s:0.0
> sd7.kr_s:0.0
> sd7._w:0.0
> sd7.r_s:0.0
> sd7.wait:0.0
> nfs4231.svc_t:0.0
> nfs4231.w_s:0.0
> nfs4231.actv:0.0
> nfs4231._b:0.0
> nfs4231.kw_s:0.0
> nfs4231.kr_s:0.0
> nfs4231._w:0.0
> nfs4231.r_s:0.0
> nfs4231.wait:0.0
> nfs4055.svc_t:0.0
> nfs4055.w_s:0.0
> nfs4055.actv:0.0
> nfs4055._b:0.0
> nfs4055.kw_s:0.0
> nfs4055.kr_s:0.0
> nfs4055._w:0.0
> nfs4055.r_s:0.0
> nfs4055.wait:0.0
> sd3.svc_t:0.0
> sd3._b:0.0
> sd3.kw_s:0.0
> sd3.wt:0.0
> sd3.w_s:0.0
> sd3._w:0.0
> sd3.r_s:0.0
> sd3.id:98.0
> sd3.wait:0.0
> sd3.sy:1.0
> sd3.us:0.0
> sd3.tin:0.0
> sd3.tout:0.0
> sd3.actv:0.0
> sd3.kr_s:0.0
> nfs4230.svc_t:0.0
> nfs4230.w_s:0.0
> nfs4230.actv:0.0
> nfs4230._b:0.0
> nfs4230.kw_s:0.0
> nfs4230.kr_s:0.0
> nfs4230._w:0.0
> nfs4230.r_s:0.0
> nfs4230.wait:0.0
> nfs4228.svc_t:0.0
> nfs4228.w_s:0.0
> nfs4228.actv:0.0
> nfs4228._b:0.0
> nfs4228.kw_s:0.0
> nfs4228.kr_s:0.0
> nfs4228._w:0.0
> nfs4228.r_s:0.0
> nfs4228.wait:0.0
> nfs4229.svc_t:0.0
> nfs4229.w_s:0.0
> nfs4229.actv:0.0
> nfs4229._b:0.0
> nfs4229.kw_s:0.0
> nfs4229.kr_s:0.0
> nfs4229._w:0.0
> nfs4229.r_s:0.0
> nfs4229.wait:0.0
> nfs1.svc_t:0.0
> nfs1.w_s:0.0
> nfs1.actv:0.0
> nfs1._b:0.0
> nfs1.kw_s:0.0
> nfs1.kr_s:0.0
> nfs1._w:0.0
> nfs1.r_s:0.0
> nfs1.wait:0.0
> nfs4212.svc_t:0.0
> nfs4212.w_s:0.0
> nfs4212.actv:0.0
> nfs4212._b:0.0
> nfs4212.kw_s:0.0
> nfs4212.kr_s:0.0
> nfs4212._w:0.0
> nfs4212.r_s:0.0
> nfs4212.wait:0.0
> ```
> 
> I could run this file *_collectd_iostat_python.py *_ without running collectd, but it cannot be loaded into collectd plugins. :(

**Comment by deniszh** [comment]
Sorry, @saina110 , I has no access to Solaris 10 sparc to test. I have access to Solaris 10 x86, and we have python plugins working there (not collectd_iostat_python.py, but other python plugins).
Try to check permissions, or move file somewhere to built-in collectd plugin directory, e.g. /opt/csw/lib/collectd/plugins/python/

**Comment by saina110** [comment]
hi deniszh, 
Thank you for your suggestions. Today I tried several ways, but still doesn't work.. Here are some ways I tried today:

> 1) Rebuild the source packages again according to the tutorial [python-plugins](https://github.com/librato/collectd-librato):

**what I did is :** 

```
./configure --prefix=/cws/yuanliao/rrdtool-1.4.4  --with-librrd=$I/cws/yuanliao/rrdtool-1.4.4     CFLAGS=-"DLT_LAZY_OR_NOW='RTLD_LAZY|RTLD_GLOBAL'"  CFLAGS="-fPIC  -D__EXTENSIONS__"  --disable-static  --disable-ascent  --disable-apple_sensors --disable-apache
```

and gmake install  --> not working

> 2) As your suggestion, I put the XX.py file under the collectd plugins directory:
> 
> ```
> <Plugin python>
>     ModulePath "/cws/yuanliao/rrdtool-1.4.4/lib/collectd/"
>     Import "collectd_iostat_python"
>     <Module collectd_iostat_python>
>         Path "/usr/bin/iostat"
>                 Interval 2
>                 Count 2
>                 Verbose False
>     </Module>
> </Plugin>
> ```
> 
> All I got is the same issue: 
> **python plugin: Found a configuration for the "collectd_iostat_python" plugin, but the plugin isn't loaded or didn't register a configuration callback.**

Because I was thinking the python plugin cannot be loaded because of the [global variable settings](http://mailman.verplant.org/pipermail/collectd/2008-March/001616.html)? , but in this case I didn't solve it :( 

Is there any method could track the issue source? If it could work, I could make several custom plugins using python which is convenient. I would be really appreciated your suggestions. Thanks in advance!

**Comment by deniszh** [comment]
Could you please put all logs which collectd produces during start?

**Comment by saina110** [comment]
Hi deniszh,

Sorry replying to you little late.  I tried several ways, but doesn't work. So I use shell scripts in exec plugin of collectd to collect the information. I think it may be the better solution for my system at this moment. In the collectd.log, it always shows the same error **python plugin: Found a configuration for the "collectd_iostat_python" plugin, but the plugin isn't loaded or didn't register a configuration callback.**  

So at this moment, I think I prefer to use shell script . Thanks deniszh your time. I am really appreciated:)
